### PR TITLE
Fix: mainloop: resolves race condition mainloop poll

### PR DESCRIPTION
**Pull request by davidvossel**
libqb requires mainloop to be able to add, remove, and the add back the same fd. This is required for libqb to be able to perform async connection authentication. Without this functionality, libqb will block during the authentication process which is won't work for us. Our implementation of mainloop has a race condition that makes this libqb functionality of poll (add, rm, add) not work properly.

Example.

- add fd - is_used = TRUE
- remove fd  - remove source from mainloop. is_used = FALSE
- add fd - is_used = TRUE
- destroy callback from previous remove sets is_used =FALSE even though the fd is actually still active.

This causes some really interesting crashes when file descriptors look like they get removed from mainloop when in fact mainloop is still monitoring them. It's possible that this race condition also causes issues with throttling events. I can't prove it, but I have a feeling we could see issues with events possibly not going out after libqb modifies the send fd to look for POLLOUT. This patch will definitely resolve that if it happens.

### CODE REVIEW

File: lib/common/mainloop.c [hunk 1]
```diff
@@ -479,18 +484,19 @@ gio_poll_dispatch_update(enum qb_loop_priority p, int32_t fd, int32_t evts,
         return -ENOMEM;
     }
 
-    /* Because unlike the poll() API, glib doesn't tell us about HUPs by default */
-    evts |= (G_IO_HUP | G_IO_NVAL | G_IO_ERR);
-
-    if (adaptor->is_used) {
+    if (adaptor->source) {
         g_source_remove(adaptor->source);
+        adaptor->source = 0;
```
Comment (beekhof) [inline lib/common/mainloop.c]: Do we need to set to 0 before calling remove?
ie.

        int tmp = adaptor->source;
        adaptor->source = 0;
        g_source_remove(tmp);

Comment (davidvossel) [inline lib/common/mainloop.c]: it shouldn't make a difference.

Comment (beekhof) [inline lib/common/mainloop.c]: In theory :)
I know I've had to do this to fix callback loops in the past.

Comment (davidvossel) [inline lib/common/mainloop.c]: yeah, i know what you mean. this shouldn't be one of those instances.

**Pull request merged**